udtube/data/conllu.py

```diff
@@ -31,47 +116,41 @@ class TokenList(collections.UserList):
     optional associated metadata.
 
     Args:
-        tokens (Iterable[Dict[str, str]]): List of tokens.
+        tokens (Iterable[Token]): iterable of tokens.
         metadata (Dict[str, Optional[str]], optional): ordered dictionary of
             string/key pairs.
     """
 
     metadata: Dict[str, Optional[str]]
 
-    def __init__(self, tokens: Iterable[Dict[str, str]], metadata=None):
+    def __init__(self, tokens: Iterable[Token], metadata=None):
         super().__init__(tokens)
         self.metadata = metadata if metadata is not None else {}
 
-    def serialize(self) -> str:
-        """Serializes the TokenList."""
+    def __str__(self) -> str:
         line_buf = []
         for key, value in self.metadata.items():
             if value:
                 line_buf.append(f"# {key} = {value}")
             else:  # `newpar` etc.
                 line_buf.append(f"# {key}")
         for token in self:
-            col_buf = []
-            for key in _fieldnames:
-                col_buf.append(token.get(key, "_"))
-            line_buf.append("\t".join(str(cell) for cell in col_buf))
+            line_buf.append(str(token))
         return "\n".join(line_buf) + "\n"
 
+    def get_tokens(self) -> List[str]:
+        """List of tokens to be fed into tokenizer."""
+        return [token.form for token in self if not token.is_mwe]
+
 
 # Parsing.
 
 
-def _maybe_parse_metadata(line: str) -> Optional[Tuple[str, Optional[str]]]:
+def _maybe_parse_metadata(line: str) -> Optional[Tuple[str, str]]:
     """Attempts to parse the line as metadata."""
     # The first group is the key; the optional third element is the value.
-    mtch = re.fullmatch(r"#\s+(.+?)(\s+=\s+(.*))?", line)
-    if mtch:
-        return mtch.group(1), mtch.group(3)
-
-
-def _parse_token(line: str) -> Dict[str, str]:
-    """Parses the line as a token."""
-    return dict(zip(_fieldnames, line.split("\t")))
+    if match := re.fullmatch(r"#\s+(.+?)(\s+=\s+(.*))?", line):
```

Comment:
beautiful

Reply:
yeah you okay with this construct? It's valid in Python 3.8 onwards (and we support 3.9 onward).

Reply:
Yep, perfectly fine. Stealing this for myself in the future, wasn't aware of this new operator.